src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/outputMonitor.ts

```diff
@@ -54,21 +64,303 @@ export class OutputMonitor extends Disposable implements IOutputMonitor {
+			return raceResult.result as IRacePollingOrPromptResult;
+		} else if (raceResult.type === 'prompt') {
+			const promptResult = raceResult.result as boolean;
+			if (promptResult) {
+				return await this._pollForOutputAndIdle(execution, true, token, languageModelsService, taskService, pollFn);
+			} else {
+				return originalResult;
+			}
+		}
+		return await pollFn();
+	}
+
+
+	private async _promptForMorePolling(command: string, token: CancellationToken, context: IToolInvocationContext, chatService: IChatService): Promise<{ promise: Promise<boolean>; part?: ChatElicitationRequestPart }> {
+		if (token.isCancellationRequested) {
+			return { promise: Promise.resolve(false) };
+		}
+		const chatModel = chatService.getSession(context.sessionId);
+		if (chatModel instanceof ChatModel) {
+			const request = chatModel.getRequests().at(-1);
+			if (request) {
+				let part: ChatElicitationRequestPart | undefined = undefined;
+				const promise = new Promise<boolean>(resolve => {
+					const thePart = part = new ChatElicitationRequestPart(
+						new MarkdownString(localize('poll.terminal.waiting', "Continue waiting for \`{0}\`?", command)),
+						new MarkdownString(localize('poll.terminal.polling', "This will continue to poll for output to determine when the terminal becomes idle for up to 2 minutes.")),
+						'',
+						localize('poll.terminal.accept', 'Yes'),
+						localize('poll.terminal.reject', 'No'),
+						async () => {
+							thePart.state = 'accepted';
+							thePart.hide();
+							thePart.dispose();
+							resolve(true);
+						},
+						async () => {
+							thePart.state = 'rejected';
+							thePart.hide();
+							resolve(true);
+						}
+					);
+					this._register(thePart);
+					chatModel.acceptResponseProgress(request, thePart);
+				});
+
+				return { promise, part };
+			}
+		}
+		return { promise: Promise.resolve(false) };
+	}
+
+	private async _pollForOutputAndIdle(
+		execution: IExecution,
+		extendedPolling: boolean,
+		token: CancellationToken,
+		languageModelsService: Pick<ILanguageModelsService, 'selectLanguageModels' | 'sendChatRequest'>,
+		taskService: ITaskService,
+		pollFn?: (execution: IExecution, token: CancellationToken, terminalExecutionIdleBeforeTimeout: boolean, pollStartTime: number, extendedPolling: boolean, languageModelsService: Pick<ILanguageModelsService, 'selectLanguageModels' | 'sendChatRequest'>, taskService: ITaskService) => Promise<IPollingResult | undefined> | undefined
+	): Promise<IPollingResult> {
+		const maxWaitMs = extendedPolling ? PollingConsts.ExtendedPollingMaxDuration : PollingConsts.FirstPollingMaxDuration;
+		const maxInterval = PollingConsts.MaxPollingIntervalDuration;
+		let currentInterval = PollingConsts.MinPollingDuration;
+		const pollStartTime = Date.now();
+
+		let lastBufferLength = 0;
+		let noNewDataCount = 0;
+		let buffer = '';
+		let terminalExecutionIdleBeforeTimeout = false;
+
+		while (true) {
+			if (token.isCancellationRequested) {
+				break;
+			}
```

Comment:
Can a generic polling function be pulled out? We want to remove this complexity around polling, so the logic is more clear

Reply:
Would `waitForIdle` be useful to simplify this?

Reply:
Can we do this:


```ts
let state = 'polling';

while (state === 'polling') {
	await waitForIdle(1000);
	const result = checkCondition();
	state = result.blah;
	// switch if state becomes complex
}
```

Example of how you could await multiple promises:

https://github.com/microsoft/vscode/blob/dfcd8f473b5c7a1f250623d1f94d86dc98c1dc06/src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/executeStrategy/basicExecuteStrategy.ts#L52-L71